### PR TITLE
test symboo.defTree for symbols from Tasty

### PR DESCRIPTION
**Pull request by liufengyun**
test symboo.defTree for symbols from Tasty

### CODE REVIEW

**Comment by liufengyun** [comment]
@nicolasstucki Now this is rebased and simplified as discussed in the meeting.

File: compiler/src/dotty/tools/dotc/core/Symbols.scala [hunk 1]
```diff
@@ -540,6 +540,11 @@ object Symbols {
       d != null && d.flagsUNSAFE.is(Private)
     }
 
+    /** Is the symbol a pattern bound symbol?
+     */
+    final def isPatternBound(implicit ctx: Context): Boolean =
+      !isClass && this.is(Case, butNot = Enum | Module)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/core/Symbols.scala]:
```suggestion
      !isClass && is(Case, butNot = Enum | Module)
```

File: compiler/src/dotty/tools/dotc/plugins/Plugin.scala [hunk 2]
```diff
@@ -28,6 +28,8 @@ sealed trait Plugin {
    *
    *  Research plugin receives a phase plan and return a new phase plan, while
    *  non-research plugin returns a list of phases to be inserted.
+   *
+   *  @note Research plugins only compile with nightly build.
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/plugins/Plugin.scala]: Do we test this?

File: sbt-dotty/sbt-test/sbt-dotty/analyzer-plugin/lib/Lib.scala [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package lib
+
+class A(val n: Int) {
+  def this(c: Char) = this(c.toInt)
+
+  val a = 30 * n
+  // val p = Product("x", 100)
```
Comment (nicolasstucki) [inline sbt-dotty/sbt-test/sbt-dotty/analyzer-plugin/lib/Lib.scala]:
```suggestion
```

File: tests/plugins/custom/analyzer/Lib_2.scala [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package lib
+
+class A(val n: Int) {
+  def this(c: Char) = this(c.toInt)
+
+  val a = 30 * n
+  // val p = Product("x", 100)
```
Comment (nicolasstucki) [inline tests/plugins/custom/analyzer/Lib_2.scala]:
```suggestion
```

File: compiler/src/dotty/tools/dotc/core/Symbols.scala (hunk 1 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/core/Symbols.scala]: Initially, I tried `is(...)`, but it does not check --- it makes sense, as adaptation only happens with a select.

File: compiler/src/dotty/tools/dotc/plugins/Plugin.scala (hunk 2 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/plugins/Plugin.scala]: The logic is enforced with the following property:

```Scala
object Properties {
  val experimental = versionString.contains("SNAPSHOT") || versionString.contains("NIGHTLY")
}
```

We have tests for research plugins. However, non-support of research plugins is only reviewed.

File: compiler/src/dotty/tools/dotc/core/Symbols.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/Symbols.scala]: Ok

**Pull request merged**